Improve wasm calli thunk lookup

radekdoulik opened the pull request:
Put the thunks into hash table, this will also serve later for the generated thunk lookups

Add CallFunc_F64_RetF64

radekdoulik committed:
Improve wasm calli thunk lookup

Put the thunks into hash table, this will also serve later for the generated thunk lookups

Add CallFunc_F64_RetF64

radekdoulik committed:
Merge branch 'main' into clr-interp-wasm-thunks-in-hash-table

radekdoulik committed:
Fix after merge

radekdoulik committed:
Fix return values

radekdoulik commented on src/coreclr/vm/wasm/helpers.cpp:
```diff
@@ -639,97 +636,127 @@ namespace
+            argType = sig.NextArg())
+        {
+            _ASSERTE(pos < maxSize);
+            keyBuffer[pos++] = GetTypeCode(ConvertibleTo(argType, sig, false /* isReturn */));
```
It should already fail in `GetTypeCode`, actually we don't need the check for NotConvertible above.

radekdoulik commented:
> > this will also serve later for the generated thunk lookups
> 
> Does Mono build the hashtable of the thunks during startup, or is the hashtable pre-built at build time?

mono currently uses binary search in static table

> We may want the latter, especially for AOT.

I think we can improve it gradually? I remember you told me that part of runtime already does something similar, we didn't talk about the details though.

We can use perfect hash or just precompute the hash values for the signature string keys. Or do you have another idea?

pavelsavara commented:
> runtime already does something similar

This is android for pinvoke by symbol name. They pre-build the hash table.
https://github.com/dotnet/android/blob/main/src/native/clr/pinvoke-override/pinvoke-tables.include

AaronRobinsonMSFT commented on src/coreclr/vm/wasm/helpers.cpp:
```diff
@@ -639,97 +636,127 @@ namespace
-        if (isVoidReturn)
+        uint32_t pos = 0;
+
+        if (sig.IsReturnTypeVoid())
+            keyBuffer[pos++] = 'v';
+        else
```
```suggestion
        if (sig.IsReturnTypeVoid())
        {
            keyBuffer[pos++] = 'v';
        }
        else
```

radekdoulik committed:
Apply suggestions from code review

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

radekdoulik committed:
Feedback

Plus removed final from the StringWasmThunkSHashTraits as
NoRemoveSHashTraits needs to derive from it

jkotas committed:
Apply suggestion from @jkotas

jkotas committed:
Apply suggestion from @jkotas
